contracts/src/main/kotlin/com/contractsAndStates/states/RoadState.kt

```diff
@@ -45,3 +45,153 @@ data class RoadState(
+    return assignLongestRoad(currentHolder, candidates)
+}
+
+data class LongestRoadCandidate(val player: Party, val longestRoadLength: Int)
+
+/**
+ * Given a list of pairs of parties and count of the longest road and the current player with
+ * the longest road card (null means no one has the card), return the new card holder. It can
+ * be a new player, the same player or null if no one meets the criteria.
+ */
+fun assignLongestRoad(holder: Party?, candidates: List<LongestRoadCandidate>): Party? {
+
+    val orderedCandidates = candidates.sortedByDescending { it.longestRoadLength }
+
+    // Longest candidate does not have at least 5 roads
+    if (orderedCandidates[0].longestRoadLength < 5)
+        return null
+
+    // Only one player has the longest road
+    if (orderedCandidates[0].longestRoadLength > orderedCandidates[1].longestRoadLength)
+        return orderedCandidates[0].player
+
+    val holderLongestRoadLength = orderedCandidates.first { it.player == holder }.longestRoadLength
+
+    // Holder and another player has same length
+    if (holderLongestRoadLength == orderedCandidates[0].longestRoadLength)
+        return holder
+
+    // All other cases. More than one player has same length and longer than current holder
+    return null
+}
+
+/**
+ * Returns a list with the count of the longest road for each player.
+ */
+fun longestRoadsForAllPlayers(board: PlacedHexTiles,
+                              roads: List<RoadState>,
+                              settlements: List<SettlementState>,
+                              players: List<Party>): List<LongestRoadCandidate> {
+    return players.map { LongestRoadCandidate(it, longestRoadForPlayer(board, roads, settlements, it).count()) }
+}
+
+/**
+ * Returns a set of road ids with the longest road for a particular player.
+ */
+fun longestRoadForPlayer(board: PlacedHexTiles,
+                         roads: List<RoadState>,
+                         settlements: List<SettlementState>,
+                         player: Party): Set<UniqueIdentifier> {
+    val playerRoadIds = roads.filter { it.owner == player }.toSet()
+    val othersSettlements = settlements.filter { it.owner != player }.toSet()
+    return calculateLongestRoad(board, playerRoadIds, othersSettlements)
+}
+
+/**
+ * Calculate the longest road for a player.
+ * roads must only contain roads for the player.
+ * settlements must only contain settlements of all the other players.
+ */
+private fun calculateLongestRoad(board: PlacedHexTiles,
+                                 roads: Set<RoadState>,
+                                 settlements: Set<SettlementState>): Set<UniqueIdentifier> {
+    val roadIds = roads.map { it.linearId }.toSet()
+    val settlementsCorners = settlements.map { AbsoluteCorner(it.hexTileIndex, it.hexTileCoordinate) }.toSet()
+
+    var longestRoad = setOf<UniqueIdentifier>()
+    val neverVisitedRoads = roads.map { it.linearId }.toMutableSet()
```

Comment:
`= roadIds.toMutableSet()`.

Reply:
I just need the ids. Not the whole RoadState

Reply:
What I mean is that `roads.map { it.linearId }` is already calculated on line 121.